### PR TITLE
Ask for email unless --allow-unsafe-registration

### LINKED ISSUE #414: Client should strongly encourage use of an email address
If the CA has many (/any) users with no email on file, it's going to cause a lot of hassles. We're still deciding if we want it to be official policy that everyone has to provide an email, but in the meantime we can work on the client side of this. Proposal: The interactive version of the client should ask for an email, and not proceed without one, unless it was invoked with a --allow-unsafe-registration. The help documentation for --allow-unsafe-registration would say:

"Specifying this flag enables registering an account with no email address. This is strongly discouraged, because in the event of key loss or account compromise you will irrevocably lose access to your account. You will also be unable to receive notice about impending expiration of revocation of your certificates. Updates to the Subscriber Agreement will still affect you, and the CA will make a best-effort attempt to notify you about them based on information about the affected domains, such as WHOIS data and common administrative addresses."

### PR DESCRIPTION
**Pull request by miquelruiz**
Add new option that explicitly allows to not provide an email.
Fixes #414

### CODE REVIEW

File: letsencrypt/cli.py [hunk 1]
```diff
@@ -127,9 +127,9 @@ def _determine_account(args, config):
         elif len(accounts) == 1:
             acc = accounts[0]
         else:  # no account registered yet
-            if args.email is None:
+            if args.email is None and not args.allow_unsafe_registration:
                 args.email = display_ops.get_email()
-            if not args.email:  # get_email might return ""
```
Comment (kuba) [inline letsencrypt/cli.py]: i think that removing this line is no good - see the comment

File: letsencrypt/cli.py [hunk 2]
```diff
@@ -127,9 +127,9 @@ def _determine_account(args, config):
         elif len(accounts) == 1:
             acc = accounts[0]
         else:  # no account registered yet
-            if args.email is None:
+            if args.email is None and not args.allow_unsafe_registration:
```
Comment (kuba) [inline letsencrypt/cli.py]: so basically `--alow-unsafe-registration` will never prompt for email? that doesn't sound good

Comment (pde) [inline letsencrypt/cli.py]: I think this is fine, though perhaps the flag should have a more accurate name, such as `--register-unsafely-without-email`.  The purpose of the flag is either use by folks who've decided they really don't want to provide an email, or automation in environments contact-based recovery is not in use.

**Comment by pde** [comment]
Thanks for the PR Miquel!  It looks like there are some unit test failures to be fixed.

Also, the ncurses UI looks a little awkward with the double parentheses and variable spacing; perhaps leave the previous question and parentheses then two newlines and say "If you really want to skip this, run the client with --register-unsafely-without-email".

**Comment by miquelruiz** [comment]
I've renamed the option and added some checks to ensure it exits if no email is provided.
I've run the unit tests in master and they seem to be failing too. Is there anything I can do?
I could try to fix them, but doesn't look like something easy for me.
.

**Comment by pde** [comment]
One reason the tests are failing is that the following command is failling:

```
sudo letsencrypt -t --email foo@gmail.com --agree-dev-preview -vvvv --debug -d example.com
```
I get:
```
...
Error: No email was provided and --register-unsafely-without-email was not present.
2015-10-29 19:10:06,697:DEBUG:letsencrypt.cli:Exiting abnormally:
Traceback (most recent call last):
  File "venv/bin/letsencrypt", line 9, in <module>
    load_entry_point('letsencrypt==0.1.0.dev0', 'console_scripts', 'letsencrypt')()
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/letsencrypt/cli.py", line 1113, in main
    return args.func(args, config, plugins)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/letsencrypt/cli.py", line 434, in run
    le_client = _init_le_client(args, config, authenticator, installer)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/letsencrypt/cli.py", line 160, in _init_le_client
    acc, acme = _determine_account(args, config)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/letsencrypt/cli.py", line 151, in _determine_account
    "Unable to register an account with ACME server")
Error: Unable to register an account with ACME server
...
```

**Comment by bmw** [comment]
Updating title with `[needs revision]`. Please remove when this is ready for review @miquelruiz.

**Comment by pde** [comment]
This has now become #1524.

**Pull request closed without merging**